core/src/main/java/io/questdb/griffin/engine/join/HashOuterJoinFilteredLightRecordCursorFactory.java

```diff
@@ -217,59 +460,136 @@ public boolean hasNext() {
         }
 
         @Override
-        public long preComputedStateSize() {
-            return isMapBuilt ? 1 : 0;
+        public void toTop() {
+            super.toTop();
+            filter.toTop();
         }
 
         @Override
-        public long size() {
-            return -1L;
+        protected void of(RecordCursor masterCursor, RecordCursor slaveCursor, SqlExecutionContext sqlExecutionContext) throws SqlException {
+            super.of(masterCursor, slaveCursor, sqlExecutionContext);
+            record.of(masterRecord, slaveRecord);
+            filter.init(this, sqlExecutionContext);
+        }
+    }
+
+    private class HashRightOuterJoinLightLightRecordCursor extends AbstractHashOuterJoinLightRecordCursor {
+        private final Map matchIdsMap;
+        private final RightOuterJoinRecord record;
+        private MapRecordCursor mapCursor;
+
+        public HashRightOuterJoinLightLightRecordCursor(
+                int columnSplit,
+                Record nullRecord,
+                Map joinKeyMap,
+                Map matchIdsMap,
+                LongChain slaveChain
+        ) {
+            super(columnSplit, joinKeyMap, slaveChain);
+            record = new RightOuterJoinRecord(columnSplit, nullRecord);
+            this.matchIdsMap = matchIdsMap;
+            isOpen = true;
         }
 
         @Override
-        public void toTop() {
-            filter.toTop();
-            masterCursor.toTop();
-            slaveChainCursor = null;
-            if (!isMapBuilt) {
-                slaveCursor.toTop();
-                joinKeyMap.clear();
-                slaveChain.clear();
+        public void close() {
+            if (isOpen) {
+                matchIdsMap.close();
             }
+            super.close();
+            mapCursor = Misc.free(mapCursor);
         }
 
-        private void buildMapOfSlaveRecords() {
-            final Record record = slaveCursor.getRecord();
-            while (slaveCursor.hasNext()) {
-                circuitBreaker.statefulThrowExceptionIfTripped();
+        @Override
+        public Record getRecord() {
+            return record;
+        }
 
+        @Override
+        public boolean hasNext() {
+            if (!isMapBuilt) {
+                populateRowIDHashMap(circuitBreaker, slaveCursor, joinKeyMap, slaveKeySink, slaveChain);
+                isMapBuilt = true;
+                record.hasMaster(true);
+                mapCursor = joinKeyMap.getCursor();
+            }
+
+            if (slaveChainCursor != null && slaveChainCursor.hasNext()) {
+                while (slaveChainCursor.hasNext()) {
```

Comment:
We've just called `slaveChainCursor.hasNext()` in the above `if`. Why are we calling it second time here?

Reply:
ditto